### PR TITLE
providers/aws: S3 bucket website support

### PR DESCRIPTION
**Pull request by justincampbell**
* [x] Create/update/destroy
* [x] index/error document
* [x] Computed endpoint
* [x] Multi-region support
* [x] Docs

```go
provider "aws" { region = "us-east-1" }

resource "aws_s3_bucket" "default" {
  bucket = "tf.justincampbel.me"
  acl = "public-read"

  website {
    index_document = "index.html"
    error_document = "error.html"
  }
}

output "dns_name" { value = "${aws_s3_bucket.default.website_endpoint}" }
// "tf.justincampbell.me.s3-website-us-east-1.amazonaws.com"
```

### CODE REVIEW

File: builtin/providers/aws/resource_aws_s3_bucket.go [hunk 1]
```diff
@@ -94,6 +124,15 @@ func resourceAwsS3BucketRead(d *schema.ResourceData, meta interface{}) error {
...
 	}
 
+	// Add website_endpoint as an output
+	endpoint, err := websiteEndpoint(s3conn, d)
+	if err != nil {
+		return err
+	}
+	if err := d.Set("website_endpoint", endpoint); err != nil {
+		return err
+	}
```
Comment (justincampbell) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: This doesn't seem to set `website_endpoint` to `""` if you disable website support on an existing bucket. How do I trigger other resources that depend on that attribute to be updated?

File: builtin/providers/aws/resource_aws_s3_bucket.go [hunk 2]
```diff
@@ -118,3 +157,73 @@ func resourceAwsS3BucketDelete(d *schema.ResourceData, meta interface{}) error {
...
+		log.Printf("[DEBUG] S3 put bucket website: %s", putInput)
+
+		_, err := s3conn.PutBucketWebsite(putInput)
+		if err != nil {
+			return fmt.Errorf("Error putting S3 website: %s", err)
+		}
+	} else {
+		deleteInput := &s3.DeleteBucketWebsiteInput{Bucket: aws.String(bucket)}
+
+		log.Printf("[DEBUG] S3 delete bucket website: %s", deleteInput)
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: I'm not sure what this should output here, something about the website? Or the bucket name? Right now for me it outputs `[DEBUG] S3 delete bucket website: &{%!s(*string=0xc20834f270) {%!s(bool=false)}}`

File: builtin/providers/aws/resource_aws_s3_bucket.go [hunk 3]
```diff
@@ -31,6 +31,31 @@ func resourceAwsS3Bucket() *schema.Resource {
...
+			"index_document": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
+			},
+
+			"error_document": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: I think `ForceNew: false,` is safe to omit; it's implied then

File: builtin/providers/aws/resource_aws_s3_bucket.go [hunk 4]
```diff
@@ -118,3 +157,73 @@ func resourceAwsS3BucketDelete(d *schema.ResourceData, meta interface{}) error {
 	}
 	return nil
 }
+
+func updateWebsite(s3conn *s3.S3, d *schema.ResourceData) error {
+	website := d.Get("website").(bool)
+	bucket := d.Get("bucket").(string)
+	indexDocument := d.Get("index_document").(string)
+	errorDocument := d.Get("error_document").(string)
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: There is a method to check for changes, `d.HasChange`, which can help you here. As it stands now, every update will do `PutBucketWebsiteInput` if `website=true`. If `website=false`, you still call `s3conn.DeleteBucketWebsite()` every time. Ex, you create a s3 bucket but do not specify `website=true`, that API is still called.

**Comment by catsby** [comment]
I think a `website {}` block would be a better UX here. Happy to pair on this if you'd like

File: builtin/providers/aws/resource_aws_s3_bucket.go [hunk 5]
```diff
@@ -118,3 +156,96 @@ func resourceAwsS3BucketDelete(d *schema.ResourceData, meta interface{}) error {
...
+	}
+
+	// Default to us-east-1 if the bucket doesn't have a region
+	if region == "" {
+		region = "us-east-1"
+	}
+
+	endpoint = fmt.Sprintf("%s.s3-website-%s.amazonaws.com", bucket, region)
+
+	return
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: This is just a personal preference, but I prefer explicit returns when possible. I don't think that's a requirement, but I believe most of TF is consistent there.

File: builtin/providers/aws/resource_aws_s3_bucket.go [hunk 6]
```diff
@@ -118,3 +156,96 @@ func resourceAwsS3BucketDelete(d *schema.ResourceData, meta interface{}) error {
...
+	if err != nil {
+		return "", err
+	}
+	var region string
+	if location.LocationConstraint != nil {
+		region = *location.LocationConstraint
+	}
+
+	// Default to us-east-1 if the bucket doesn't have a region
+	if region == "" {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: Will AWS every actually return a non-nil value that's an empty string?

**Comment by catsby** [comment]
Looks solid! :+1:

File: builtin/providers/aws/resource_aws_s3_bucket.go (hunk 6 above)
Comment (justincampbell) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: Added that link to the comment

**Comment by justincampbell** [comment]
@catsby do these look ok? http://cl.ly/image/0F03161P2o0S/Screen%20Shot%202015-05-01%20at%209.53.04%20AM.png

**Comment by catsby** [comment]
Are we OK with this PR as is, with #1769 being a separate PR?

**Comment by mitchellh** [comment]
It looks like the `Read` fucntion doesn't update the website index/error document. Can we get that fixed? Other than that it looks good!

File: builtin/providers/aws/resource_aws_s3_bucket.go (hunk 1 above)
Comment (phinze) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: I'd actually just add an acceptance test that includes the disable case to test this. So it'd be a test with one step that has website support and the next step that removes it.

Comment (justincampbell) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: Done in bb63f5c

**Comment by justincampbell** [comment]
@mitchellh done!

**Comment by phinze** [comment]
Alright this looks ready to :ship: :ok_woman:

**Pull request merged**